merkle/smt/hstar3.go

```diff
@@ -70,32 +70,43 @@ func NewHStar3(updates []NodeUpdate, hash HashChildrenFn, depth, top uint) (HSta
 	return HStar3{upd: updates, hash: hash, depth: depth, top: top}, nil
 }
 
-// Prepare returns the set of all the node IDs that the Update method will load
-// in order to compute node hash updates from the initial tree depth up to the
-// top level specified in the constructor. It may be useful for constructing a
-// NodeAccessor, e.g. by batch-reading the nodes from elsewhere.
-//
-// Note: The returned map could have bool value type, but []byte allows the
-// caller to reuse this map for filling in the hashes for the Update method.
+// Prepare returns the list of all the node IDs that the Update method will
+// load in order to compute node hash updates from the initial tree depth up to
+// the top level specified in the constructor. It may be useful for creating a
+// a NodeAccessor, e.g. by batch-reading the nodes from elsewhere.
 //
 // TODO(pavelkalinnikov): Return only tile IDs.
-func (h HStar3) Prepare() map[tree.NodeID2][]byte {
-	ids := make(map[tree.NodeID2][]byte)
+func (h HStar3) Prepare() []tree.NodeID2 {
+	empty := tree.NodeID2{}
+	ids := []tree.NodeID2{empty} // Start with a dummy sentinel value.
```

Comment:
May be worth considering initializing `ids` with a capacity >>> 1. In the worst case you need `len(h.upd) * (h.depth-h.top)` elements. May be worth pre-allocating that?

Reply:
Yes! This decreased the benchmark running time from 30ms to 6ms.